notification/slack/usergrouperror.go

```diff
@@ -0,0 +1,82 @@
+)
+
+// userGroupErrorMissing is a template for when a user-group update fails because one or more users are missing from Slack.
+var userGroupErrorMissing = template.Must(template.New("userGroupErrorMissing").Parse(`Hey everyone! I couldn't update {{.GroupRef}} because I couldn't find the following user(s) in Slack: {{.MissingUserRefs}}
+
+If you could have them click the button below to connect their Slack account, that would be great! Hopefully I'll be able to update the user-group next time.
+
+{{.LinkedUserRefs}}`))
+
+// userGroupErrorEmpty is a template for when a user-group update fails because there are no users on-call.
+var userGroupErrorEmpty = template.Must(template.New("userGroupErrorMissing").Parse(`Hey everyone! I couldn't update {{.GroupRef}} because there is nobody on-call for {{.ScheduleRef}}.
+
+Since a Slack user-group cannot be empty, I'm going to leave it as-is for now.`))
+
+// userGroupErrorUpdate is a template for when a user-group update fails for any reason other than missing users.
+var userGroupErrorUpdate = template.Must(template.New("userGroupErrorMissing").Parse(`Hey everyone! I couldn't update {{.GroupRef}} because I ran into a problem. Maybe touch base with the GoAlert admin(s) to see if they can help? I'm sorry for the inconvenience!
+
+Here's the ID I left with the error in my logs so they can find it:
+{{.ErrorRef}}`))
+
+// userGroupError has information about an error that occurred while updating a user-group.
+type userGroupError struct {
+	ErrorID      uuid.UUID
+	GroupID      string
+	ScheduleID   string
+	ScheduleName string
+	Missing      []notification.User
+	Linked       []string
+
+	callbackFunc func(string, ...url.Values) string
+}
+
+// ErrorRef returns a string that can be used to reference the error in a Slack message.
+func (e userGroupError) ErrorRef() string {
+	return fmt.Sprintf("`SlackUGErrorID=%s`", e.ErrorID.String())
+}
+
+// GroupRef returns a string that can be used to reference the user-group in a Slack message.
+func (e userGroupError) GroupRef() string {
+	return fmt.Sprintf("<!subteam^%s>", e.GroupID)
+}
+
+// MissingUserRefs returns a string that can be used to reference the missing users in a Slack message.
+func (e userGroupError) MissingUserRefs() string {
+	var refs []string
+	for _, u := range e.Missing {
+		urlStr := e.callbackFunc(fmt.Sprintf("users/%s", url.PathEscape(u.ID)))
+		refs = append(refs, slackLink(urlStr, u.Name))
+	}
+	return strings.Join(refs, ", ")
+}
+
+// LinkedUserRefs returns a string that can be used to reference the linked users in a Slack message.
+func (e userGroupError) LinkedUserRefs() string {
+	var refs []string
+	for _, u := range e.Linked {
+		refs = append(refs, fmt.Sprintf("<@%s>", u))
+	}
+	return strings.Join(refs, " ")
+}
+
+// ScheduleRef returns a string that can be used to reference the schedule in a Slack message.
+func (e userGroupError) ScheduleRef() string {
+	urlStr := e.callbackFunc("schedules/" + url.PathEscape(e.ScheduleID))
```

Comment:
Using `Sprintf` is no more or less secure, though it does add unnecessary noise in this case. The important piece here is using `PathEscape`.

Reply:
Based on corporate secure code training, directly concatenating strings across all languages is directly called out as either outright dangerous, or a poor habit.  While you're right in that since golang is a typed language that solves a lot of these issues, we still want to set a good example to reinforce good secure coding habits.